### DISCUSSION TITLE
Can jfreechart 1.5.3 be used with jlink to create a self contained executable?

### DISCUSSION THREAD

**Discussion by AlexHomebrew** [jfree/jfreechart#374] [Q&A] [answered]
I am adding charts to a large legacy project created in Netbeans and chose jfreechart. I have already migrated the project from Java 8 to JDK 15.0.2 and Netbeans 15, which included modularization. I added jlink to create a self contained executable so my customers don't have to install java. So far, all my installations are in Windows 10.

When I compile (in Netbeans using Maven) I get the following error:
Error: automatic module cannot be used with jlink: org.jfree.jfreechart from file:///C:/Users/AlexXX/.m2/repository/org/jfree/jfreechart/1.5.3/jfreechart-1.5.3.jar
Command execution failed.
org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1)
    at org.apache.commons.exec.DefaultExecutor.executeInternal (DefaultExecutor.java:404)
. . .

Is there a way to modularize jfreechart so it doesn't rely on being an automatic module?
Or more to the point, can I create an executable that includes everything it needs (like jlink creates) with jfreechart?
Or, is JFreeChart 2.0 coming out soon?
Thanks,
Alex

**Comment by trashgod** [comment]
[JFreeChart](https://github.com/jfree/jfreechart) version 2.0.0 is modular but not yet released. I haven't tried `jlink` with it, but you should be able to build, install and require it like [this](https://stackoverflow.com/a/76276885/230513). Alternatively, [JFree Demos](https://github.com/jfree/jfree-demos) illustrates using the [`maven-assembly-plugin`](https://maven.apache.org/plugins/maven-assembly-plugin/) to create a JAR with dependencies.

**Comment by AlexHomebrew** [comment, marked as answer]
I created a Hack that allowed me to build with jlink and include JFreeChart-1.5.3.
1. Download the source code into a new Netbeans project called `JFreeChart`.
2. Create a new file:  `JFreeChart/src/main/java/module-info.java` containing the following
```
module jfreechart
{
    requires java.desktop;
    requires java.logging;
    requires java.sql;

    exports org.jfree.chart;
    exports org.jfree.chart.axis;
    exports org.jfree.chart.plot;
    exports org.jfree.chart.renderer.xy;
    exports org.jfree.chart.ui;
    exports org.jfree.data.time;
}
```
3. Remove the package (and contents) `org.jfree.chart.servlet`
4. Add `requires jfreechart;` to my project's `module-info.java`

JLink now is happy and builds the project, which runs. I still have to verify that I can create a chart and display it.
Note that my project doesn't need the servlet functionality and I removed it because the `servlet.api` is not modularized so jlink was unhappy. Fortunately, it was self contained (as good designs should be) so I didn't need to modify anything else. As I build more charts, I may need to add more `exports`, but that is easy.

I will report back after I create some working charts.

**Reply by trashgod** [reply]
Will this be a configuration management problem going forward? Version 2 already elides servlets, and `jlink` should optimize the rest. For reference, this [gist](https://gist.github.com/trashgod/a3f7581417db19bf5b6dc52d3647a2f9) illustrates migrating a Swing application; see tag `v2modular`.

**Reply by AlexHomebrew** [reply]
You are correct @trashgod  it will, as most hacks are. But I need to move forward now and I am expecting to have some other changes when upgrading to JFreeChart 2.0. I am documenting what I did to minimize that effort. Thanks for your feedback.

**Comment by AlexHomebrew** [comment]
My chart is displaying great (thank you to David Gilbert)
